pkg/pcr/replay_test.go

```diff
@@ -0,0 +1,122 @@
+package pcr
+
+import (
+	"testing"
+
+	"github.com/9elements/converged-security-suite/v2/pkg/tpmeventlog"
+	"github.com/google/go-tpm/tpm2"
+	"github.com/stretchr/testify/require"
+)
+
+func TestReplay(t *testing.T) {
+	makeMeasurementDigest := func(alg tpm2.Algorithm, lastByteValue uint8) []byte {
+		h, err := alg.Hash()
+		require.NoError(t, err)
+		r := make([]byte, h.Size())
+		r[len(r)-1] = lastByteValue
+		return r
+	}
+
+	makeFinalDigest := func(alg tpm2.Algorithm, lastByteValues []uint8) []byte {
+		h, err := alg.Hash()
+		require.NoError(t, err)
+		final := make([]byte, h.Size())
+		final[len(final)-1] = lastByteValues[0]
+		extend := func(d []byte) {
+			hasher := h.HashFunc().New()
+			_, err = hasher.Write(final)
+			require.NoError(t, err)
+			_, err = hasher.Write(d)
+			require.NoError(t, err)
+			final = hasher.Sum(nil)
+		}
+		single := make([]byte, h.Size())
+		for _, lastByteValue := range lastByteValues[1:] {
+			single[len(single)-1] = lastByteValue
+			extend(single)
+		}
+		return final
+	}
+
+	eventLog := &tpmeventlog.TPMEventLog{
+		Events: []*tpmeventlog.Event{
+			{
+				PCRIndex: 0,
+				Type:     tpmeventlog.EV_NO_ACTION,
+				Data:     []byte("StartupLocality\x00\x01"), // TPM init locality is 1
+				Digest: &tpmeventlog.Digest{
+					HashAlgo: tpm2.AlgSHA1,
+					Digest:   makeMeasurementDigest(tpm2.AlgSHA1, 2),
+				},
+			},
+			{
+				PCRIndex: 0,
+				Type:     tpmeventlog.EV_NO_ACTION,
+				Digest: &tpmeventlog.Digest{
+					HashAlgo: tpm2.AlgSHA256,
+					Digest:   makeMeasurementDigest(tpm2.AlgSHA256, 3),
+				},
+			},
+			{
+				PCRIndex: 0,
+				Type:     tpmeventlog.EV_S_CRTM_CONTENTS,
+				Digest: &tpmeventlog.Digest{
+					HashAlgo: tpm2.AlgSHA1,
+					Digest:   makeMeasurementDigest(tpm2.AlgSHA1, 4),
+				},
+			},
+			{
+				PCRIndex: 0,
+				Type:     tpmeventlog.EV_S_CRTM_CONTENTS,
+				Digest: &tpmeventlog.Digest{
+					HashAlgo: tpm2.AlgSHA256,
+					Digest:   makeMeasurementDigest(tpm2.AlgSHA256, 5),
+				},
+			},
+			{
+				PCRIndex: 1,
+				Type:     tpmeventlog.EV_S_CRTM_CONTENTS,
+				Digest: &tpmeventlog.Digest{
+					HashAlgo: tpm2.AlgSHA1,
+					Digest:   makeMeasurementDigest(tpm2.AlgSHA1, 6),
+				},
+			},
+			{
+				PCRIndex: 1,
+				Type:     tpmeventlog.EV_EFI_VARIABLE_BOOT,
+				Digest: &tpmeventlog.Digest{
+					HashAlgo: tpm2.AlgSHA1,
+					Digest:   makeMeasurementDigest(tpm2.AlgSHA1, 7),
+				},
+			},
+		},
+	}
+
+	t.Run("positive", func(t *testing.T) {
```

Comment:
Comment on what exactly we're testing here would be nice.

Reply:
Fixed.